examples/water/main.rs

```diff
@@ -0,0 +1,797 @@
+            },
+            sample_count: 1,
+            sample_mask: !0,
+            alpha_to_coverage_enabled: false,
+        });
+
+        // Done
+        let this = Example {
+            water_vertex_buf,
+            water_vertex_count: water_vertices.len(),
+            water_bind_group,
+            water_uniform_buf,
+            water_pipeline,
+
+            terrain_vertex_buf,
+            terrain_vertex_count: terrain_vertices.len(),
+            terrain_normal_bind_group,
+            terrain_flipped_bind_group,
+            terrain_normal_uniform_buf,
+            terrain_flipped_uniform_buf,
+            terrain_pipeline,
+
+            reflect_view: reflection_texture.create_default_view(),
+
+            draw_depth_buffer: draw_depth_buffer.create_default_view(),
+
+            current_frame: 0,
+
+            active: 0,
+        };
+        (this, None)
+    }
+
+    fn update(&mut self, _event: winit::event::WindowEvent) {
+        //empty
+    }
+
+    fn resize(
+        &mut self,
+        sc_desc: &wgpu::SwapChainDescriptor,
+        device: &wgpu::Device,
+        queue: &wgpu::Queue,
+    ) {
+        if sc_desc.width == 0 && sc_desc.height == 0 {
+            // Stop rendering altogether.
+            self.active = std::usize::MAX;
+            return;
+        } else {
+            // Skip a frame, since it crashes otherwise.
+            // The next frame queued is the wrong size (0, 0).
+            self.active = self.current_frame + 1;
+        }
+
+        // Regenerate all of the buffers and textures.
+        let texture_extent = wgpu::Extent3d {
+            width: sc_desc.width,
+            height: sc_desc.height,
+            depth: 1,
+        };
+
+        let reflection_texture = device.create_texture(&wgpu::TextureDescriptor {
+            label: None,
+            size: texture_extent,
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: sc_desc.format,
+            usage: wgpu::TextureUsage::SAMPLED | wgpu::TextureUsage::COPY_DST | wgpu::TextureUsage::OUTPUT_ATTACHMENT,
+        });
+
+        let draw_depth_buffer = device.create_texture(&wgpu::TextureDescriptor {
+            label: None,
+            size: texture_extent,
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Depth32Float,
+            usage: wgpu::TextureUsage::SAMPLED | wgpu::TextureUsage::COPY_DST | wgpu::TextureUsage::OUTPUT_ATTACHMENT,
+        });
+
+        self.draw_depth_buffer = draw_depth_buffer.create_default_view();
+        self.reflect_view = reflection_texture.create_default_view();
+
+        let (
+            normal_view,
+            flipped_view,
+            projection,
+        ) = Self::generate_matrices(sc_desc.width as f32 / sc_desc.height as f32);
+
+        let normal_view: &[f32; 16] = normal_view.as_ref();
```

Comment:
let's put all of this initialization code in a static method that we can re-use on both initialization and resize